design_drafts/async_controller_impl.md

```diff
@@ -0,0 +1,40 @@
+## Implementation of Asynchronous Controllers
+
+
+#### Short description
+
+Asynchronous controllers are an upcoming addition to the ros2_control framework. As the name suggests, they make it possible to run the update calls of these controllers on a separate thread. Asynchronous components are omitted here for the sake of clarity, but they follow a very similar structure.
+Since the intention was to support real-time systems, the standard library's std::atomic interface was used, and due to the inherent difficulty of writing and understanding concurrent lock-free programs, a brief glance at the code might not be enough to figure out if something doesn't work as intended.
+
+For that reason, the purpose of this design draft is to explain how the synchronization works between the Controller Manager's thread and an asynchronous controllers, now to make it easier for others to review the PR of the feature, and later to let the users know if the current implementation is suitable for their use case.
+
+#### Implementation
+
+The controllers running asynchronously are entirely independent of each other, which means that no data sharing occurs between them. Therefore synchronization happens between the Controller Manager's thread and an async controller - so in the following drawings only these two threads will be visualized.
+
+The critical sections are:
+
+* The state interface writes of the Controller Manager's read function
+* The command interface writes of the async controller's update function
+
+The former has to occur and be seen by the async thread before the update, while the latter, ideally, has to finish and be seen before the controller manager's write.
+
+![async_controller](https://user-images.githubusercontent.com/25421074/212979058-5dad88c1-d940-4bc7-ac6f-103711de7210.png)
+
+As shown in the above image, there are four different cases to account for, and these are: 
+
+1. The state interfaces aren't written yet by the cm read. In this case, we can be sure that the release flag isn't set, thus the acquire load on the other side will return false - so we'll skip the update and the asynchronous controller will sleep until the next iteration. This works as intended.
+
+2. The state interfaces are written in the cm read function, and the release flag is set. It can happen that the state interfaces are visible on the controller's side, but the flag isn't - in this case, we'll do the same as the first scenario. On the other hand, if the flag is set, then the thread's acquire load will return true if it isn't currently sleeping, and the respective orderings guarantee that the async update will be able to work with the latest state interface values. This is fine too.
```

Comment:
I am uncertain if I understand who is releasing and how is acquiring stuff in this case. This general explanation would be good.

Reply:
Not sure I understand the second sentence

Reply:
I've created a easy to understand C++ example with some explanations.